### ISSUE TITLE
recv failed : Connection reset by peer？

### ISSUE THREAD

**Issue by Sampson1107** [uber/horovod#147]
mpirun --allow-run-as-root  -np 4 -H localhost:4 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -mca pml ob1 -mca btl_openib_receive_queues P,128,32:P,2048,32:P,12288,32:P,65536,32 python scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --model resnet101 --batch_size 64 --variable_update horovod

node1:54573:54696 [1] include/socket.h:194 WARN Call to recv failed : Connection reset by peer
node1:54573:54696 [1] INFO transport/net_ib.cu:537 -> 2
node1:54573:54696 [1] INFO transport/net_ib.cu:610 -> 2
node1:54573:54696 [1] INFO include/net.h:30 -> 2 [Net]
node1:54573:54696 [1] INFO include/net.h:38 -> 2 [Net]
node1:54573:54696 [1] INFO bootstrap.cu:169 -> 2
node1:54573:54696 [1] INFO init.cu:400 -> 2
node1:54573:54696 [1] INFO init.cu:517 -> 2

**Comment by alsrgv** [comment]
@Sampson1107, 

What messages were in the log prior to `TENCENT64:16736:16986 [1] include/socket.h:194 WARN Call to recv failed : Connection reset by peer`?

Can you check `ifconfig`, and if you see interfaces like `docker0`, add `-x NCCL_SOCKET_IFNAME=^docker0`?

**Comment by Sampson1107** [comment]
@alsrgv There are no docker0
eth0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether 24:8a:07:e1:9b:80  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 100.102.32.175  netmask 255.255.255.128  broadcast 100.102.32.255
        ether 24:8a:07:e1:9b:81  txqueuelen 1000  (Ethernet)
        RX packets 25690211399  bytes 35138998855198 (31.9 TiB)
        RX errors 0  dropped 2009131  overruns 2009131  frame 0
        TX packets 22239239636  bytes 29569512302428 (26.8 TiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 0  (Local Loopback)
        RX packets 951298965  bytes 27297574733533 (24.8 TiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 951298965  bytes 27297574733533 (24.8 TiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

**Comment by Sampson1107** [comment]
@alsrgv 
There is no docker0.
eth0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
......

eth1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
  ......

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
......

**Comment by Sampson1107** [comment]
@alsrgv 
The log like this:
Extracting MNIST-data-0/train-images-idx3-ubyte.gz
Extracting MNIST-data-1/train-images-idx3-ubyte.gz
Extracting MNIST-data-1/train-labels-idx1-ubyte.gz
Extracting MNIST-data-1/t10k-images-idx3-ubyte.gz
Extracting MNIST-data-0/train-labels-idx1-ubyte.gz
Extracting MNIST-data-0/t10k-images-idx3-ubyte.gz
Extracting MNIST-data-1/t10k-labels-idx1-ubyte.gz
Extracting MNIST-data-0/t10k-labels-idx1-ubyte.gz
WARNING:tensorflow:From tensorflow_mnist.py:84: get_or_create_global_step (from tensorflow.contrib.framework.python.ops.variables) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.get_or_create_global_step
WARNING:tensorflow:From tensorflow_mnist.py:84: get_or_create_global_step (from tensorflow.contrib.framework.python.ops.variables) is deprecated and will be removed in a future version.
Instructions for updating:
Please switch to tf.train.get_or_create_global_step
INFO:tensorflow:Create CheckpointSaverHook.
2018-01-16 14:29:26.534912: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-01-16 14:29:26.604305: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2018-01-16 14:29:27.368731: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties: 
name: Tesla M40 24GB major: 5 minor: 2 memoryClockRate(GHz): 1.112
pciBusID: 0000:05:00.0
totalMemory: 23.90GiB freeMemory: 23.78GiB
2018-01-16 14:29:27.368772: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 1, name: Tesla M40 24GB, pci bus id: 0000:05:00.0, compute capability: 5.2)
2018-01-16 14:29:27.472391: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties: 
name: Tesla M40 24GB major: 5 minor: 2 memoryClockRate(GHz): 1.112
pciBusID: 0000:04:00.0
totalMemory: 23.90GiB freeMemory: 23.78GiB
2018-01-16 14:29:27.472432: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: Tesla M40 24GB, pci bus id: 0000:04:00.0, compute capability: 5.2)
node1:54572:54698 [0] INFO NET : Using interface eth1:100.102.32.175<0>
node1:54572:54698 [0] INFO NET/IB : Using interface eth1 for sideband communication
node1:54572:54698 [0] INFO NET/IB: [0] mlx4_0:2/RoCE 
node1:54572:54698 [0] INFO Using internal Network IB
node1:54572:54698 [0] INFO Using NCCL Low-latency algorithm for sizes below 16384
NCCL version 2.1.2+cuda8.0
node1:54573:54696 [1] INFO NET : Using interface eth1:100.102.32.175<0>
node1:54573:54696 [1] INFO NET/IB : Using interface eth1 for sideband communication
node1:54573:54696 [1] INFO NET/IB: [0] mlx4_0:2/RoCE 
node1:54573:54696 [1] INFO Using internal Network IB
node1:54573:54696 [1] INFO Using NCCL Low-latency algorithm for sizes below 16384


node1:54572:54963 [0] transport/net_ib.cu:757 WARN NET/IB : Got completion with error 12, opcode 0, vendor err 129
node1:54572:54963 [0] INFO include/net.h:33 -> 1 [Net]
node1:54572:54963 [0] INFO include/net.h:47 -> 2 [Net]
node1:54572:54963 [0] INFO bootstrap.cu:63 -> 2 [bthread]

node1:54573:54696 [1] include/socket.h:194 WARN Call to recv failed : Connection reset by peer
node1:54573:54696 [1] INFO transport/net_ib.cu:537 -> 2
node1:54573:54696 [1] INFO transport/net_ib.cu:610 -> 2
node1:54573:54696 [1] INFO include/net.h:30 -> 2 [Net]
node1:54573:54696 [1] INFO include/net.h:38 -> 2 [Net]
node1:54573:54696 [1] INFO bootstrap.cu:169 -> 2
node1:54573:54696 [1] INFO init.cu:400 -> 2
node1:54573:54696 [1] INFO init.cu:517 -> 2

**Comment by alsrgv** [comment]
@Sampson1107, looks like something is happening with your `mlx4_0` interface:
```
node1:54572:54963 [0] transport/net_ib.cu:757 WARN NET/IB : Got completion with error 12, opcode 0, vendor err 129
```

Can you try to add `-x NCCL_IB_DISABLE=1` and see if it helps?

**Comment by Sampson1107** [comment]
@alsrgv 
Thanks， it works!
Will this decrease the performance of the training speed?

**Comment by alsrgv** [comment]
@Sampson1107, if you will run Horovod across multiple machines, RoCE should give you better performance.  In that case, we'd have to figure out why NCCL is failing to initialize it.

**Comment by Sampson1107** [comment]
@alsrgv 
Alsrgv, I will run Horovod across multiple machines and I want to get a high training speed. Is there any bug or is it my problem?

**Comment by Sampson1107** [comment]
The nccl version I used  is nccl_2.1.2  and ROCE is used for my server. The mlx4_0 is as following(the port 1 is down):
CA 'mlx4_0'
	CA type: MT4103
	Number of ports: 2
	Firmware version: 2.36.5000
	Hardware version: 0
	Node GUID: 0x248a07030068f380
	System image GUID: 0x248a07030068f380
	Port 1:
		State: Down
		Physical state: Disabled
		Rate: 40
		Base lid: 0
		LMC: 0
		SM lid: 0
		Capability mask: 0x04010000
		Port GUID: 0x268a07fffe68f380
		Link layer: Ethernet
	Port 2:
		State: Active
		Physical state: LinkUp
		Rate: 40
		Base lid: 0
		LMC: 0
		SM lid: 0
		Capability mask: 0x04010000
		Port GUID: 0x248a07000168f381
		Link layer: Ethernet

**Comment by alsrgv** [comment]
According to quick googling:
```
Status 12 = IB_WC_RETRY_EXC_ERR
Vendor_err = 129 --> Timeout and transport error counter exceeded

This indicates that we lost connection to the client ie. something went
wrong on client side (bad operation cause QP error...) please try to
catch any error on the client (qp async event, cq error status and
vendor_err...)

Today I just run vdbench on big file and get error right away (lost
connection and nfsrdma cannot recover from there)
```
(from http://lists.openfabrics.org/pipermail/ewg/2010-February/014826.html)

@bureddy, do you have any pointers or user guides for @Sampson1107 to diagnose this problem?

**Comment by Sampson1107** [comment]
@alsrgv 
Thanks a lot! Uninstall the IB driver and the problem is solved.

**Comment by alsrgv** [comment]
Removing the driver will have the same effect as NCCL_IB_DISABLE. It would be good to figure out source of the error and fix it to get a better perf.

**Comment by Sampson1107** [comment]
@alsrgv 
OK. Perhaps it has two network drivers. I will figure out the reason the next step!

**Comment by EdwardZhang88** [comment]
Hi @Sampson1107 I have the same error as 'transport/net_ib.cu:781 WARN NET/IB : Got completion with error 12, opcode 1, len0, vendor err 129' although I am running the mpi task in a docker container. I also tried adding option and env variable like OMPI_MCA_btl=^openib and NCCL_IB_HCA=mlx5_0 but nothing really worked. Till now, I have to either uninstall Mallanox drive or set NCCL_IB_DISABLE=1 as you did above to get it run through. I am wondering if you are now able to use RoCE rather than TCP socket when doing distributed training across nodes? If so, can you please shed some light here? Thanks.
